From 3703a0478b09471a2ea2f68786f88beb316eebd7 Mon Sep 17 00:00:00 2001
From: rohit <rohit.sharma@r2p.com>
Date: Wed, 8 Nov 2023 11:14:41 +0000
Subject: [PATCH] fix device tree for olimex a20 som carrier board

Configures the A20 SOM Carrier Board with:
- Additional UART ports (uart2, uart3, uart4)
- Audio output via TPA2012D2 amplifier on PH15 (enable) and PH14 (gain)
- Microphone input via internal codec
- CSI0 camera interface
- Removes HDMI (not present on carrier board)
---
 arch/arm/boot/dts/sun7i-a20-olinuxino-micro.dts | 113 ++++++++++++++++++------
 1 file changed, 91 insertions(+), 22 deletions(-)

diff --git a/arch/arm/boot/dts/sun7i-a20-olinuxino-micro.dts b/arch/arm/boot/dts/sun7i-a20-olinuxino-micro.dts
index 359bd0d5b3b1..a1b2c3d4e5f6 100644
--- a/arch/arm/boot/dts/sun7i-a20-olinuxino-micro.dts
+++ b/arch/arm/boot/dts/sun7i-a20-olinuxino-micro.dts
@@ -57,7 +57,10 @@
 	aliases {
 		serial0 = &uart0;
 		serial1 = &uart6;
-		serial2 = &uart7;
+		serial2 = &uart2;
+		serial3 = &uart4;
+		serial4 = &uart3;
+		serial5 = &uart7;
 		spi0 = &spi1;
 		spi1 = &spi2;
 	};
@@ -66,17 +69,6 @@
 		stdout-path = "serial0:115200n8";
 	};

-	hdmi-connector {
-		compatible = "hdmi-connector";
-		type = "a";
-
-		port {
-			hdmi_con_in: endpoint {
-				remote-endpoint = <&hdmi_out_con>;
-			};
-		};
-	};
-
 	leds {
 		compatible = "gpio-leds";
 		pinctrl-names = "default";
@@ -87,6 +79,51 @@
 			gpios = <&pio 7 2 GPIO_ACTIVE_HIGH>;
 			default-state = "on";
 		};
+	};
+
+	speaker_amp: speaker-amp {
+		compatible = "simple-audio-amplifier";
+		/* PH15: HIGH=shutdown, LOW=enabled */
+		enable-gpios = <&pio 7 15 GPIO_ACTIVE_LOW>;
+		sound-name-prefix = "Speaker Amp";
+	};
+
+	sound {
+		compatible = "simple-audio-card";
+		simple-audio-card,name = "A20-SOM-Carrier";
+
+		simple-audio-card,aux-devs = <&speaker_amp>;
+
+		simple-audio-card,widgets =
+			"Speaker", "Speaker",
+			"Microphone", "Microphone";
+
+		simple-audio-card,routing =
+			"Speaker Amp INL", "HP Left",
+			"Speaker Amp INR", "HP Right",
+			"Speaker", "Speaker Amp OUTL",
+			"Speaker", "Speaker Amp OUTR",
+			"MIC1", "Microphone";
+
+		simple-audio-card,cpu {
+			sound-dai = <&codec>;
+		};
+
+		simple-audio-card,codec {
+			sound-dai = <&codec>;
+		};
+	};
+
+	audio_gain: audio-gain {
+		compatible = "regulator-fixed";
+		regulator-name = "audio-gain";
+		/* PH14 controls GAIN1 pin of TPA2012D2 */
+		gpio = <&pio 7 14 GPIO_ACTIVE_HIGH>;
+		regulator-min-microvolt = <3300000>;
+		regulator-max-microvolt = <3300000>;
+		enable-active-high;
+		regulator-always-on;
+		status = "okay";
 	};
 };

@@ -96,6 +133,14 @@
 };

 &codec {
+	#sound-dai-cells = <0>;
+	pinctrl-names = "default";
+	pinctrl-0 = <&audio_pins>;
+	allwinner,audio-routing =
+		"HP Left", "HPOUTL",
+		"HP Right", "HPOUTR",
+		"MIC1", "Mic Bias",
+		"Mic Bias", "Microphone";
 	status = "okay";
 };

@@ -123,16 +168,6 @@
 	status = "okay";
 };

-&hdmi {
-	status = "okay";
-};
-
-&hdmi_out {
-	hdmi_out_con: endpoint {
-		remote-endpoint = <&hdmi_con_in>;
-	};
-};
-
 &i2c0 {
 	status = "okay";

@@ -254,6 +289,17 @@
 		function = "gpio_out";
 		drive-strength = <20>;
 	};
+
+	audio_pins: audio-pins {
+		pins = "PH14", "PH15";
+		function = "gpio_out";
+		drive-strength = <20>;
+	};
+
+	csi0_8bits_pins: csi0-8bits-pins {
+		pins = "PE0", "PE1", "PE2", "PE3", "PE4", "PE5", "PE6", "PE7", "PE8", "PE9", "PE10", "PE11";
+		function = "csi0";
+	};
 };

 #include "axp209.dtsi"
@@ -323,6 +369,26 @@
 	status = "okay";
 };

+&uart2 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&uart2_pi_pins>, <&uart2_cts_rts_pi_pins>;
+	uart-has-rtscts;
+	status = "okay";
+};
+
+&uart3 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&uart3_pg_pins>, <&uart3_cts_rts_pg_pins>;
+	uart-has-rtscts;
+	status = "okay";
+};
+
+&uart4 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&uart4_pg_pins>;
+	status = "okay";
+};
+
 &uart6 {
 	pinctrl-names = "default";
 	pinctrl-0 = <&uart6_pi_pins>;
@@ -335,6 +401,12 @@
 	status = "okay";
 };

+&csi0 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&csi0_8bits_pins>;
+	status = "okay";
+};
+
 &usb_otg {
 	dr_mode = "otg";
 	status = "okay";
--
2.42.1

